### ISSUE TITLE
Implement Port Control Protocol

### ISSUE THREAD

**Issue by DentonGentry** [tailscale/tailscale#2563]
Seen in logs for an unrelated issue:

```
portmapper: unexpected PMP response: 02 80
portmapper: unexpected PMP response: 02 81 00 02 a2 a9 00 00 00 4f 8f d2
```

This happens when parsePMPResponse returns !ok.

```
func parsePMPResponse(pkt []byte) (res pmpResponse, ok bool) {
        if len(pkt) < 12 {
                return
        }
        ver := pkt[0]
        if ver != 0 {
                return
        }
```

The RFC for NAT-PMP https://datatracker.ietf.org/doc/html/rfc6886#section-1.1 says, _"The version number in the packet header is 0 for NAT-PMP and 2 for PCP, so the packets are easily distinguished."_

So this is the Port Control Protocol being seen in the wild.

**Comment by bradfitz** [comment]
Which version logged this?

We know PCP exists in the wild, but it was so rarely the only option that we haven't implemented it (despite being easy). But now that we do more than 1, adding a 3rd should be easy. (@JulianKnodt, interested?)

I'm surprised at this log message, though. Is a PMP query resulting in a PCP response?

**Comment by DentonGentry** [comment]
This came from 1.12.1 on an iOS device. The user reported a different problem, which I'm looking into, and happened to notice this.

It does appear that we sent a NAT-PMP request and we got back a PCP response.

**Comment by JulianKnodt** [comment]
I can try it out next week! altho that one I'd need to dig thru the spec which hopefully will be short

**Comment by JulianKnodt** [comment]
This issue should be closed after 1.14 when this is included

**Comment by DentonGentry** [comment]
Fixed in 1.14.
